Package request: pytorch on darwin with GPU (MPS) support

**Project description**

Not sure if this is a "packaging request" but didn't seem like a bug report, just wanted a tracking issue, I might try to hack on this eventually.

Pytorch on aarch64-darwin now supports GPU:

https://pytorch.org/docs/stable/notes/mps.html

The [currently packaged pytorch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/torch/default.nix) runs but doesn't use my Mac's GPU:

```
$ nix-shell -p 'python3.withPackages (ps: with ps; [ pytorch ])' --command "python -c 'import torch; print(torch.backends.mps.is_available())'"
False
```

In contrast, using the pytorch-provided wheel:

```console
$ python3 -m venv .venv && source ./.venv/bin/activate
$ pip install https://files.pythonhosted.org/packages/85/68/f901437d3e3ef6fe97adb1f372479626d994185b8fa06803f5bdf3bb90fd/torch-2.0.1-cp311-none-macosx_11_0_arm64.whl
$ python -c 'import torch; print(torch.backends.mps.is_available())'
True
```

**Metadata**

* homepage URL: https://pytorch.org/
* source URL: https://github.com/pytorch/pytorch
* license: bsd3 https://github.com/pytorch/pytorch/blob/main/LICENSE
* platforms: darwin

cc @NixOS/darwin-maintainers

Just for tests I also tried to use apple-sdk_12 but I got these errors:
```
python3.12-torch>   return MTLLanguageVersion3_0;
python3.12-torch>          ^~~~~~~~~~~~~~~~~~~~~
python3.12-torch>          MTLLanguageVersion2_0
python3.12-torch> /nix/store/2vs624s5vsdnhralhwlin5gnid2dl1qy-apple-sdk-12.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Metal.framework/Headers/MTLLibrary.h:185:5: note: 'MTLLanguageVersion2_0' declared here
python3.12-torch>     MTLLanguageVersion2_0 API_AVAILABLE(macos(10.13), ios(11.0)) = (2 << 16),
python3.12-torch>     ^
python3.12-torch> 1 error generated.
python3.12-torch> [2278/5844] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/api/src/serialize/output-archive.cpp.o
python3.12-torch> ninja: build stopped: subcommand failed.
python3.12-torch> ERROR Backend subprocess exited when trying to invoke build_wheel
error: builder for '/nix/store/n41ysm7qfk9azq8w09vs5l8iwvmwqvr3-python3.12-torch-2.5.0.drv' failed with exit code 1;
       last 10 log lines:
       >          ^~~~~~~~~~~~~~~~~~~~~
       >          MTLLanguageVersion2_0
       > /nix/store/2vs624s5vsdnhralhwlin5gnid2dl1qy-apple-sdk-12.3/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Metal.framework/Headers/MTLLibrary.h:185:5: note: 'MTLLanguageVersion2_0' declared here
       >     MTLLanguageVersion2_0 API_AVAILABLE(macos(10.13), ios(11.0)) = (2 << 16),
       >     ^
       > 1 error generated.
       > [2278/5844] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/torch/csrc/api/src/serialize/output-archive.cpp.o
       > ninja: build stopped: subcommand failed.
       >
       > ERROR Backend subprocess exited when trying to invoke build_wheel
       For full logs, run 'nix log /nix/store/n41ysm7qfk9azq8w09vs5l8iwvmwqvr3-python3.12-torch-2.5.0.drv'.
```

So I guess 13 is way to go

--- a/pkgs/development/python-modules/torch/default.nix
+++ b/pkgs/development/python-modules/torch/default.nix
@@ -35,7 +35,7 @@
   removeReferencesTo,
 
   # Build inputs
-  darwin,
+  apple-sdk_13,
   numactl,
 
   # dependencies
@@ -364,6 +364,9 @@ buildPythonPackage rec {
   # NB technical debt: building without NNPACK as workaround for missing `six`
   USE_NNPACK = 0;
 
+  # Explicitly enable MPS for Darwin
+  USE_MPS = setBool stdenv.hostPlatform.isDarwin;
+
   cmakeFlags =
     [
       # (lib.cmakeBool "CMAKE_FIND_DEBUG_MODE" true)
@@ -519,9 +522,7 @@ buildPythonPackage rec {
     ++ lib.optionals (cudaSupport || rocmSupport) [ effectiveMagma ]
     ++ lib.optionals stdenv.hostPlatform.isLinux [ numactl ]
     ++ lib.optionals stdenv.hostPlatform.isDarwin [
-      darwin.apple_sdk.frameworks.Accelerate
-      darwin.apple_sdk.frameworks.CoreServices
-      darwin.libobjc
+      apple-sdk_13
     ]
     ++ lib.optionals tritonSupport [ _tritonEffective ]
     ++ lib.optionals MPISupport [ mpi ]

